ampproject/amp-wp
# Improve handling of form submissions that redirect and implement form submission proxy

## Summary

* Add support for automatically redirecting users when WordPress uses `header('Location: …')` in the same way as before `wp_redirect(…)` was able to automatically supply the `AMP-Redirect-To` response header. This should eliminate the need for [this shim code](https://gist.github.com/westonruter/b2554073461707a88057bb86b6b603d6#file-amp-gravity-form-submission-shim-php-L12-L29) in Gravity Forms. The corresponding [playbook information](https://amp-wp.org/documentation/playbooks/handling-form-submissions/) will also be obsolete with this PR.
* Add support for submitting forms to external domains by means of a new REST API form submission proxy. 

Fixes #4191.

Iterates on #2425. Revisits #911. See also #923.

## Testing

1. First of all, fire up [this Glitch to be able to test remote form submissions](https://glitch.com/~brook-paranthodon-k6ttvyw9a).
2. Then install the [Test POST Form](https://gist.github.com/westonruter/6c0eb253972ccdfc31f841629e2f4344) plugin.
3. Add two shortcodes to a post: `[test_post_form]` and `[test_post_form action="https://brook-paranthodon-k6ttvyw9a.glitch.me/"]`. For example:
```html

<p><strong>Internal action:</strong></p>



[test_post_form]



<p><strong>External action:</strong></p>



[test_post_form action="https://brook-paranthodon-k6ttvyw9a.glitch.me/?foo=bar"]

```
4. Check the behavior of submitting the internal action and external action forms when on an AMP page.

Any submission request type that has a JSON response (including the use of `wp_die()`) should have a response looking like:

![image](https://user-images.githubusercontent.com/134745/73888591-ce923e00-4822-11ea-8924-2319c29cbd63.png)

or 

![image](https://user-images.githubusercontent.com/134745/73888701-f5507480-4822-11ea-9ab4-9a9e033864af.png)

If the form handler does _not_ respond with something that can be represented as JSON (e.g. returning with an HTML page) then the expected output should look like:

![image](https://user-images.githubusercontent.com/134745/73888779-1add7e00-4823-11ea-890e-391c85a7818b.png)

or 

![image](https://user-images.githubusercontent.com/134745/73888806-2a5cc700-4823-11ea-8af2-3017031b90da.png)

Note that the underlying HTTP status code and status text should be displayed in this message.

Additionally, if a form submission results in a redirect, either via `wp_redirect()` or using an underlying `header('Location: …')` call, then the user should be redirected after seeing a “Redirecting…” message momentarily:

![image](https://user-images.githubusercontent.com/134745/73889156-e0281580-4823-11ea-83d5-abc93fe1b2f8.png)

Similar behavior should work both for when the form submission is handled internally by the present WordPress or if the form submission is pointing to another domain.

## Checklist

- [ ] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).
